{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "CAAAGTCGGTA\n",
      "AAAGT\n",
      "AAGTC\n",
      "AGTCG\n",
      "CAAAG\n",
      "CGGTA\n",
      "GTCGG\n",
      "TCGGT\n"
     ]
    }
   ],
   "source": [
    "def allKmerLexicographically(k, s):\n",
    "    '''\n",
    "    s : Dna String\n",
    "    k : length of k-mer\n",
    "    '''\n",
    "    \n",
    "    l = len(s)\n",
    "\n",
    "    k_mer_list = []\n",
    "\n",
    "    for i in range(l + 1 - k):\n",
    "        k_mer_list.append(s[i:i+k])\n",
    "    k_mer_list.sort()\n",
    "    \n",
    "    return k_mer_list\n",
    "\n",
    "\n",
    "k = int(input())\n",
    "s = input()\n",
    "\n",
    "k_mer_list = allKmerLexicographically(k, s)\n",
    "\n",
    "for i in k_mer_list:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCGAAGCT\n"
     ]
    }
   ],
   "source": [
    "def stringFromGenomePath(genome_path):\n",
    "    '''\n",
    "    genome_path : DNA list\n",
    "    '''\n",
    "    \n",
    "    s = genome_path[0][:-1]\n",
    "    for i in genome_path:\n",
    "        s += i[-1]\n",
    "    return s\n",
    "\n",
    "genome_path = [\"ACCGA\",\n",
    "               \"CCGAA\",\n",
    "               \"CGAAG\",\n",
    "               \"GAAGC\",\n",
    "               \"AAGCT\"]\n",
    "print(stringFromGenomePath(genome_path))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGGCA -> GGCAT\n",
      "CATGC -> ATGCG\n",
      "GCATG -> CATGC\n",
      "GGCAT -> GCATG\n"
     ]
    }
   ],
   "source": [
    "def overlapGraph(pattern_list):\n",
    "    '''\n",
    "    pattern_list : DNA list\n",
    "    '''\n",
    "    \n",
    "    adjacency_list = dict()\n",
    "    \n",
    "    k = len(pattern_list[0])\n",
    "    \n",
    "    for i in pattern_list:\n",
    "        suf_i = i[1:k]\n",
    "        \n",
    "        for j in pattern_list:\n",
    "            if j != i:\n",
    "                pre_j = j[0:k-1]\n",
    "                if suf_i == pre_j:\n",
    "                    if i in adjacency_list and not j in adjacency_list[i]:\n",
    "                        adjacency_list[i].append(j)\n",
    "                    else:\n",
    "                        adjacency_list[i] = [j]\n",
    "    return adjacency_list\n",
    "\n",
    "pattern_list = [\"ATGCG\",\n",
    "                \"GCATG\",\n",
    "                \"CATGC\",\n",
    "                \"AGGCA\",\n",
    "                \"GGCAT\"]\n",
    "\n",
    "adjacency_list_graph = overlapGraph(pattern_list)\n",
    "\n",
    "for i in sorted(adjacency_list_graph):\n",
    "    all_prefix = \"\"\n",
    "    for j in adjacency_list_graph[i]:\n",
    "        if len(all_prefix):\n",
    "            all_prefix += (\",\" + str(j))\n",
    "        else:\n",
    "            all_prefix += (str(j))\n",
    "\n",
    "    print(str(i)+ \" -> \" +str(all_prefix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAG -> AGA\n",
      "AGA -> GAT\n",
      "ATT -> TTC\n",
      "CTA -> TAC\n",
      "CTC -> TCT\n",
      "GAT -> ATT\n",
      "TCT -> CTA,CTC\n",
      "TTC -> TCT\n"
     ]
    }
   ],
   "source": [
    "def deBruijnGraphFromString(k, s):\n",
    "    '''\n",
    "    k : length of k-mer \n",
    "    s : DNA String\n",
    "    '''\n",
    "    \n",
    "    pattern_list = allKmerLexicographically(k-1, s)\n",
    "    adjacency_list_graph = overlapGraph(pattern_list)\n",
    "    return adjacency_list_graph\n",
    "    \n",
    "k = 4\n",
    "s = \"AAGATTCTCTAC\"\n",
    "\n",
    "adjacency_list_graph = deBruijnGraphFromString(k, s)\n",
    "\n",
    "for i in sorted(adjacency_list_graph):\n",
    "    all_prefix = \"\"\n",
    "    for j in adjacency_list_graph[i]:\n",
    "        if len(all_prefix):\n",
    "            all_prefix += (\",\" + str(j))\n",
    "        else:\n",
    "            all_prefix += (str(j))\n",
    "\n",
    "    print(str(i)+ \" -> \" +str(all_prefix))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGG -> GGG\n",
      "CAG -> AGG,AGG\n",
      "GAG -> AGG\n",
      "GGA -> GAG\n",
      "GGG -> GGG,GGA\n"
     ]
    }
   ],
   "source": [
    "def deBruijnGraphFromKmerList(k_mer_list):\n",
    "    adjacency_list = dict()\n",
    "    \n",
    "    k = len(k_mer_list[0])\n",
    "    \n",
    "    for i in k_mer_list:\n",
    "        suf_i = i[1:k]\n",
    "        pre_i = i[0:k-1]\n",
    "\n",
    "        if pre_i in adjacency_list:\n",
    "            adjacency_list[pre_i].append(suf_i)\n",
    "        else:\n",
    "            adjacency_list[pre_i] = [suf_i]\n",
    "    return adjacency_list\n",
    "\n",
    "\n",
    "k_mer_list = [\"GAGG\",\n",
    "              \"CAGG\",\n",
    "              \"GGGG\",\n",
    "              \"GGGA\",\n",
    "              \"CAGG\",\n",
    "              \"AGGG\",\n",
    "              \"GGAG\"]\n",
    "\n",
    "adjacency_list_graph = deBruijnGraphFromKmerList(k_mer_list)\n",
    "\n",
    "for i in sorted(adjacency_list_graph):\n",
    "    all_prefix = \"\"\n",
    "    for j in adjacency_list_graph[i]:\n",
    "        if len(all_prefix):\n",
    "            all_prefix += (\",\" + str(j))\n",
    "        else:\n",
    "            all_prefix += (str(j))\n",
    "\n",
    "    print(str(i)+ \" -> \" +str(all_prefix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTC\n",
      "ATC\n",
      "TTC\n",
      "ATC\n",
      "TTC\n"
     ]
    }
   ],
   "source": [
    "import operator\n",
    "\n",
    "def getProfileWithPseudocount(list_s):\n",
    "    '''\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    ans = [[],[],[],[]]\n",
    "    score = 0\n",
    "    t = len(list_s)\n",
    "    k = len(list_s[0])\n",
    "    \n",
    "    for i in range(k):\n",
    "        #print(i)\n",
    "        probability = {'A':0.0, 'C':0.0, 'G':0.0, 'T':0.0}\n",
    "\n",
    "\n",
    "        for l in list_s:\n",
    "            if l[i] == 'A':\n",
    "                probability['A'] += 1.0\n",
    "            elif l[i] == 'C':\n",
    "                probability['C'] += 1.0\n",
    "            elif l[i] == 'G':\n",
    "                probability['G'] += 1.0\n",
    "            elif l[i] == 'T':\n",
    "                probability['T'] += 1.0\n",
    "        \n",
    "        probability['A'] += 1.0\n",
    "        probability['C'] += 1.0\n",
    "        probability['G'] += 1.0\n",
    "        probability['T'] += 1.0\n",
    "\n",
    "        for j in probability.keys():\n",
    "            probability[j] /= (float(t) + 4.0)\n",
    "\n",
    "        ans[0].append(probability['A'])\n",
    "        ans[1].append(probability['C'])\n",
    "        ans[2].append(probability['G'])\n",
    "        ans[3].append(probability['T'])\n",
    "        \n",
    "    return ans\n",
    "\n",
    "\n",
    "def mostProbableMotif(s, k, probabilities):\n",
    "    '''\n",
    "    s : String from which to get the bestMotif\n",
    "    k : length of k-mer\n",
    "    probabilities : profile\n",
    "    '''\n",
    "    \n",
    "    max_prob = 0.0\n",
    "    profile_most = s[0: k]\n",
    "    \n",
    "    for i in range(len(s) - k + 1):\n",
    "        sub = s[i: i + k]\n",
    "        prob = 1.0\n",
    "        for j in range(k):\n",
    "            if sub[j] == 'A':\n",
    "                prob *= probabilities[0][j]\n",
    "            elif sub[j] == 'C':\n",
    "                prob *= probabilities[1][j]\n",
    "            elif sub[j] == 'G':\n",
    "                prob *= probabilities[2][j]\n",
    "            elif sub[j] == 'T':\n",
    "                prob *= probabilities[3][j]\n",
    "\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            profile_most = sub\n",
    "\n",
    "    return profile_most\n",
    "\n",
    "\n",
    "def scoreConsensus(list_s):\n",
    "    '''\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    t = len(list_s)\n",
    "    k = len(list_s[0])\n",
    "\n",
    "    consensus = []\n",
    "    score = 0\n",
    "\n",
    "    for i in range(k):\n",
    "        #print(i)\n",
    "        frequency = dict()\n",
    "        max_freq = 0\n",
    "\n",
    "        for s in list_s:\n",
    "            if not s[i] in frequency:\n",
    "                frequency[s[i]] = 1\n",
    "            else:\n",
    "                frequency[s[i]] += 1\n",
    "            if frequency[s[i]] > max_freq:\n",
    "                max_freq = frequency[s[i]]\n",
    "                max_freq_char = s[i]\n",
    "\n",
    "        #print(frequency)\n",
    "        consensus.append(max_freq_char)\n",
    "        score += (t - max_freq)\n",
    "\n",
    "    ans = ''.join(consensus)\n",
    "    return (score, ans)\n",
    "\n",
    "\n",
    "def greedyMotifSearchWithPseudocount(k, list_s):\n",
    "    '''\n",
    "    k : length of k-mer\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    t = len(list_s)\n",
    "    lengthOfDna1 = len(list_s[0])\n",
    "    bestMotifList = []\n",
    "\n",
    "    for i in range(t):\n",
    "        bestMotifList.append(list_s[i][0:k])\n",
    "\n",
    "    bestScore, _ = scoreConsensus(bestMotifList)\n",
    "\n",
    "    for i in range(lengthOfDna1 + 1 - k):\n",
    "        sub = list_s[0][i:i + k]\n",
    "        motif_list = []\n",
    "        motif_list.append(sub)\n",
    "        profile = getProfileWithPseudocount(motif_list)\n",
    "        for j in range(t):\n",
    "            if(j != 0):\n",
    "                motif_list.append(mostProbableMotif(list_s[j], k, profile))\n",
    "                profile = getProfileWithPseudocount(motif_list)\n",
    "        score, _ = scoreConsensus(motif_list)\n",
    "        #print(motif_list, score)\n",
    "        if score < bestScore:\n",
    "            bestScore = score\n",
    "            bestMotifList = motif_list\n",
    "\n",
    "    return bestMotifList\n",
    "\n",
    "\n",
    "k = 3\n",
    "list_s = [\"GGCGTTCAGGCA\",\n",
    "     \"AAGAATCAGTCA\",\n",
    "     \"CAAGGAGTTCGC\",\n",
    "     \"CACGTCAATCAC\",\n",
    "     \"CAATAATATTCG\"]\n",
    "\n",
    "bestMotifList = greedyMotifSearchWithPseudocount(k, list_s)\n",
    "for i in bestMotifList:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtask 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCTCGGGG\n",
      "CCAAGGTG\n",
      "TACAGGCG\n",
      "TTCAGGTG\n",
      "TCCACGTG\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import operator\n",
    "\n",
    "\n",
    "def randomKmer(s, k):\n",
    "    '''\n",
    "    s : String from which to get the Kmer\n",
    "    k : length of k-mer\n",
    "    '''\n",
    "    \n",
    "    i = random.randint(0, len(s) - k)\n",
    "    return s[i: i+k]\n",
    "\n",
    "\n",
    "def randomMotifs(list_s, k):\n",
    "    '''\n",
    "    list_s : DNA list\n",
    "    k : length of k-mer\n",
    "    '''\n",
    "    \n",
    "    randomMotifList = []\n",
    "    \n",
    "    for s in list_s:\n",
    "        randomMotifList.append(randomKmer(s, k))\n",
    "        \n",
    "    return randomMotifList\n",
    "\n",
    "\n",
    "def getProfileWithPseudocount(list_s):\n",
    "    '''\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    ans = [[],[],[],[]]\n",
    "    score = 0\n",
    "    t = len(list_s)\n",
    "    k = len(list_s[0])\n",
    "    \n",
    "    for i in range(k):\n",
    "        #print(i)\n",
    "        probability = {'A':0.0, 'C':0.0, 'G':0.0, 'T':0.0}\n",
    "\n",
    "\n",
    "        for l in list_s:\n",
    "            if l[i] == 'A':\n",
    "                probability['A'] += 1.0\n",
    "            elif l[i] == 'C':\n",
    "                probability['C'] += 1.0\n",
    "            elif l[i] == 'G':\n",
    "                probability['G'] += 1.0\n",
    "            elif l[i] == 'T':\n",
    "                probability['T'] += 1.0\n",
    "        \n",
    "        probability['A'] += 1.0\n",
    "        probability['C'] += 1.0\n",
    "        probability['G'] += 1.0\n",
    "        probability['T'] += 1.0\n",
    "\n",
    "        for j in probability.keys():\n",
    "            probability[j] /= (float(t) + 4.0)\n",
    "\n",
    "        ans[0].append(probability['A'])\n",
    "        ans[1].append(probability['C'])\n",
    "        ans[2].append(probability['G'])\n",
    "        ans[3].append(probability['T'])\n",
    "        \n",
    "    return ans\n",
    "\n",
    "\n",
    "def mostProbableMotif(s, k, probabilities):\n",
    "    '''\n",
    "    s : String from which to get the bestMotif\n",
    "    k : length of k-mer\n",
    "    probabilities : profile\n",
    "    '''\n",
    "    \n",
    "    max_prob = 0.0\n",
    "    profile_most = s[0: k]\n",
    "    \n",
    "    for i in range(len(s) - k + 1):\n",
    "        sub = s[i: i + k]\n",
    "        prob = 1.0\n",
    "        for j in range(k):\n",
    "            if sub[j] == 'A':\n",
    "                prob *= probabilities[0][j]\n",
    "            elif sub[j] == 'C':\n",
    "                prob *= probabilities[1][j]\n",
    "            elif sub[j] == 'G':\n",
    "                prob *= probabilities[2][j]\n",
    "            elif sub[j] == 'T':\n",
    "                prob *= probabilities[3][j]\n",
    "\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            profile_most = sub\n",
    "\n",
    "    return profile_most\n",
    "\n",
    "\n",
    "def scoreConsensus(list_s):\n",
    "    '''\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    t = len(list_s)\n",
    "    k = len(list_s[0])\n",
    "\n",
    "    consensus = []\n",
    "    score = 0\n",
    "\n",
    "    for i in range(k):\n",
    "        #print(i)\n",
    "        frequency = dict()\n",
    "        max_freq = 0\n",
    "\n",
    "        for s in list_s:\n",
    "            if not s[i] in frequency:\n",
    "                frequency[s[i]] = 1\n",
    "            else:\n",
    "                frequency[s[i]] += 1\n",
    "            if frequency[s[i]] > max_freq:\n",
    "                max_freq = frequency[s[i]]\n",
    "                max_freq_char = s[i]\n",
    "\n",
    "        #print(frequency)\n",
    "        consensus.append(max_freq_char)\n",
    "        score += (t - max_freq)\n",
    "\n",
    "    ans = ''.join(consensus)\n",
    "    return (score, ans)\n",
    "\n",
    "\n",
    "def repeatedRandomizedMotifSearch(k, t, list_s):\n",
    "    '''\n",
    "    k : length of k-mer\n",
    "    t : total number of DNA in DNA list\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    bestScore = float('inf')\n",
    "    bestMotifList=[]\n",
    "    i = 0\n",
    "    \n",
    "    while True:\n",
    "        motifs = randomizedMotifSearch(k, t, list_s)\n",
    "        score, _ = scoreConsensus(motifs)\n",
    "        \n",
    "        if score < bestScore:\n",
    "            bestScore = score\n",
    "            bestMotifList = motifs\n",
    "            i = 0\n",
    "        else:\n",
    "            i += 1\n",
    "            \n",
    "        if i > 1000:\n",
    "            break;\n",
    "            \n",
    "    return bestMotifList\n",
    "        \n",
    "    \n",
    "def randomizedMotifSearch(k, t, list_s):\n",
    "    '''\n",
    "    k : length of k-mer\n",
    "    t : total number of DNA in DNA list\n",
    "    list_s : DNA list\n",
    "    '''\n",
    "    \n",
    "    lengthOfDna1 = len(list_s[0])\n",
    "    bestMotifList = randomMotifs(list_s, k)\n",
    "\n",
    "    bestScore, _ = scoreConsensus(bestMotifList)\n",
    "\n",
    "    while True:\n",
    "        \n",
    "        profile = getProfileWithPseudocount(bestMotifList)\n",
    "        motif_list = [mostProbableMotif(s, k, profile) for s in list_s]\n",
    "        score, _ = scoreConsensus(motif_list)\n",
    "        \n",
    "        if score < bestScore:\n",
    "            bestScore = score\n",
    "            bestMotifList = motif_list\n",
    "        else:\n",
    "            return bestMotifList\n",
    "\n",
    "\n",
    "k = 8 \n",
    "t = 5\n",
    "list_s = [\"CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA\",\n",
    "          \"GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG\",\n",
    "          \"TAGTACCGAGACCGAAAGAAGTATACAGGCGT\",\n",
    "          \"TAGATCAAGTTTCAGGTGCACGTCGGTGAACC\",\n",
    "          \"AATCCACCAGCTCCACGTGCAATGTTGGCCTA\"]\n",
    "\n",
    "bestMotifList = repeatedRandomizedMotifSearch(k, t, list_s)\n",
    "for i in bestMotifList:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda3]",
   "language": "python",
   "name": "conda-env-anaconda3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
